File: mpi/solutions/mpi_send/mpi_send.f90
PROGRAM mpi_send_example
   USE mpi_f08
   IMPLICIT NONE

   INTEGER :: rank, value, size, ierr
   TYPE(MPI_Status) :: status
   INTEGER :: input_unit, io_status

   CALL MPI_Init(ierr)
   
   CALL MPI_Comm_rank(MPI_COMM_WORLD, rank, ierr) ! get the rank or ID of the current process
   CALL MPI_Comm_size(MPI_COMM_WORLD, size, ierr) ! number of processes that are running

   ! Open input file once at the beginning (only for rank 0)
   IF (rank == 0) THEN
      OPEN(UNIT=10, FILE='input.txt', STATUS='OLD', ACTION='READ', IOSTAT=io_status)
      IF (io_status /= 0) THEN
         WRITE(*,*) 'Warning: Could not open input.txt, will exit immediately'
         input_unit = -1
      ELSE
         input_unit = 10
      END IF
   END IF

   DO
      IF (rank == 0) THEN
         WRITE(*,*) 'Enter a number to send (input negative to stop):'
         IF (input_unit > 0) THEN
            READ(input_unit, *, IOSTAT=io_status) value
            IF (io_status /= 0) THEN
               value = -1  ! Default to exit if file reading fails or EOF
            ELSE
               WRITE(*,'(A,I0)') 'Read value: ', value
            END IF
         ELSE
            value = -1  ! Default to exit if file couldn't be opened
         END IF
         CALL MPI_Send(value, 1, MPI_INTEGER, rank+1, 0, MPI_COMM_WORLD, ierr) ! the root process sends the read-in value to next process (by ID)
      ELSE
         CALL MPI_Recv(value, 1, MPI_INTEGER, rank-1, 0, MPI_COMM_WORLD, status, ierr) ! every process except the root receives the sent value from the previous process ID
         
         IF (rank < size-1) THEN
            CALL MPI_Send(value, 1, MPI_INTEGER, rank+1, 0, MPI_COMM_WORLD, ierr) ! each process sends the value to the next process by ID
         END IF
      
         WRITE(*,'(A,I0,A,I0)') 'Process ', rank, ' got ', value ! the received value
      END IF
      
      IF (value < 0) EXIT ! keep going until user inputs negative number
   END DO

   ! Close input file at the end
   IF (rank == 0 .AND. input_unit > 0) THEN
      CLOSE(input_unit)
   END IF

! Properly shutdown MPI (finalise MPI)
   CALL MPI_Finalize(ierr)

END PROGRAM mpi_send_example

File: mpi/solutions/mpi_send/mpi_send.c
#include <stdio.h>
#include "mpi.h"


////////////
//MPI_Send
////////////
//
// uses:
// int MPI_Send(const void *buf, int count, MPI_Datatype datatype, int dest, int tag, MPI_Comm comm)
//
// This example sets up a ring of processes, the user gives a value and the root process sends the value
// to the next process. Each process then sends the value to the next process (by process ID) up to the end of the ring.
//
// example usage:
//		compile: mpicc -o mpi_send mpi_send.c
//		run: mpirun -n 4 mpi_send
//
int main(argc, argv)
int argc;
char **argv;
{
    int rank, value, size;
    MPI_Status status;
    FILE *input_file = NULL;

    MPI_Init(&argc, &argv);
 
    MPI_Comm_rank(MPI_COMM_WORLD, &rank); //get the rank or ID of the current process
    MPI_Comm_size(MPI_COMM_WORLD, &size); //number of processes that are running

    // Open input file once at the beginning (only for rank 0)
    if (rank == 0) {
        input_file = fopen("input.txt", "r");
        if (!input_file) {
            printf("Warning: Could not open input.txt, will exit immediately\n");
        }
    }

    do { 
        if (rank == 0) {
            printf("Enter a number to send (input negative to stop):\n");
            if (input_file && fscanf(input_file, "%d", &value) == 1) {
                printf("Read value: %d\n", value);
            } else {
                value = -1; // Default to exit if file reading fails or EOF
            }
            MPI_Send(&value,1,MPI_INT,rank+1,0,MPI_COMM_WORLD); // the root process sends the read-in value to next process (by ID)
        } else {
            MPI_Recv(&value,1,MPI_INT,rank-1,0,MPI_COMM_WORLD,&status); // every process except to the root receives the sent value from the previous process ID
            
            if (rank < size-1)
                MPI_Send(&value,1,MPI_INT,rank+1,0,MPI_COMM_WORLD); // each process sends the value to the next process by ID
        
            printf("Process %d got %d\n",rank, value);  //the received value
        }
    } while (value >= 0); //keep going until user inputs negative number

    // Close input file at the end
    if (rank == 0 && input_file) {
        fclose(input_file);
    }

    MPI_Finalize();

    return 0;
}